File: src/main/java/rs/helpkit/internal/RSCanvas.java
package rs.helpkit.internal;

import rs.helpkit.util.fx.GraphicsState;

import java.awt.*;
import java.awt.event.*;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.function.Consumer;
import java.util.List;

/**
 * @author Tyler Sedlar
 * @since 3/20/2018
 */
public class RSCanvas extends Canvas {

    public static final Dimension GAME_SIZE = new Dimension(765, 503);

    public BufferedImage buffer, raw;

    private final Canvas original;
    private boolean normalized = false;
    private boolean hidden = false;

    public final List<Consumer<Graphics2D>> consumers = new ArrayList<>();

    public RSCanvas(Canvas original) {
        this.original = original;
        raw = new BufferedImage(GAME_SIZE.width, GAME_SIZE.height, BufferedImage.TYPE_3BYTE_BGR);
        buffer = new BufferedImage(GAME_SIZE.width, GAME_SIZE.height, BufferedImage.TYPE_3BYTE_BGR);
        for (MouseListener listener : original.getMouseListeners()) {
            original.removeMouseListener(listener);
            addMouseListener(listener);
        }
        for (MouseMotionListener listener : original.getMouseMotionListeners()) {
            original.removeMouseMotionListener(listener);
            addMouseMotionListener(listener);
        }
        for (MouseWheelListener listener : original.getMouseWheelListeners()) {
            original.removeMouseWheelListener(listener);
            addMouseWheelListener(listener);
        }
        for (KeyListener listener : original.getKeyListeners()) {
            original.removeKeyListener(listener);
            addKeyListener(listener);
        }
        for (FocusListener listener : original.getFocusListeners()) {
            original.removeFocusListener(listener);
            addFocusListener(listener);
        }
    }

    public void normalize() {
        normalized = true;
    }

    public void hideGraphics() {
        hidden = true;
    }

    @Override
    public boolean hasFocus() {
        return true;
    }

    @Override
    public Graphics getGraphics() {
        Graphics g = super.getGraphics();
        if (hidden) {
            g.setColor(getBackground());
            g.fillRect(0, 0, getWidth(), getHeight());
            return g;
        }
        if (normalized) {
            return g;
        }
        if (g != null && buffer != null && raw != null) {
            Graphics2D paint = buffer.createGraphics();
            paint.drawImage(raw, 0, 0, null);
            consumers.forEach(consumer -> {
                try {
                    GraphicsState state = new GraphicsState(paint);
                    consumer.accept(paint);
                    state.restore();
                } catch (Exception ignored) {
                    ignored.printStackTrace();
                }
            });
            paint.dispose();
            g.drawImage(buffer, 0, 0, null);
            g.dispose();
            return raw.createGraphics();
        }
        return null;
    }

    @Override
    public int hashCode() {
        return original.hashCode();
    }
}

File: src/main/java/rs/helpkit/internal/RSCanvas.kt
package rs.helpkit.internal

import rs.helpkit.OSRSContainer
import rs.helpkit.util.fx.GraphicsState
import java.awt.Canvas
import java.awt.Dimension
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.event.MouseListener
import java.awt.event.MouseMotionListener
import java.awt.event.MouseWheelListener
import java.awt.image.BufferedImage
import java.util.*
import javax.swing.event.MouseInputAdapter

/**
 * @author Tyler Sedlar
 * @since 3/20/2018
 */
class RSCanvas(container: OSRSContainer, var original: Canvas) : Canvas() {

    var buffer: BufferedImage? = null
    var raw: BufferedImage? = null

    private var normalized = false
    private var hidden = false

    private val initialMouseListeners: Array<MouseListener>
    private val initialMouseMotionListeners: Array<MouseMotionListener>
    private val initialMouseWheelListeners: Array<MouseWheelListener>
    var mouseInputAdapter: MouseInputAdapter? = null

    val consumers: MutableList<(Graphics2D) -> Unit> = ArrayList()

    init {
        bounds = original.bounds
        raw = BufferedImage(GAME_SIZE.width, GAME_SIZE.height, BufferedImage.TYPE_3BYTE_BGR)
        buffer = BufferedImage(GAME_SIZE.width, GAME_SIZE.height, BufferedImage.TYPE_3BYTE_BGR)
        initialMouseListeners = original.mouseListeners
        initialMouseMotionListeners = original.mouseMotionListeners
        initialMouseWheelListeners = original.mouseWheelListeners
        mouseInputAdapter = InputRedirector.createMouseAdapter(container, initialMouseListeners,
                initialMouseMotionListeners, initialMouseWheelListeners)
    }

    fun normalize() {
        normalized = true
    }

    fun hideGraphics() {
        hidden = true
    }

    override fun hasFocus(): Boolean {
        return true
    }

    override fun getGraphics(): Graphics? {
        val g = original.graphics
        if (hidden) {
            g!!.color = background
            g.fillRect(0, 0, width, height)
            return g
        }
        if (normalized) {
            return g
        }
        if (g != null && buffer != null && raw != null) {
            val paint = buffer!!.createGraphics()
            paint.drawImage(raw, 0, 0, null)
            consumers.forEach { consumer ->
                try {
                    val state = GraphicsState(paint)
                    consumer(paint)
                    state.restore()
                } catch (ignored: Exception) {
                    ignored.printStackTrace()
                }
            }
            paint.dispose()
            g.drawImage(buffer, 0, 0, null)
            g.dispose()
            return raw!!.createGraphics()
        }
        return null
    }

    override fun hashCode(): Int {
        return original.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        return original == other
    }

    companion object {

        val GAME_SIZE = Dimension(765, 503)
    }
}
